Compatibility with camlp5 v8.03.00

a2line opened the pull request:
So we have [failing CI jobs on linux/mac](https://github.com/geneweb/geneweb/actions/runs/9220821914/job/25368598942#step:5:51) since yesterday because of a change in camlp5 8.03, see issue https://github.com/camlp5/camlp5/issues/106.
Maintainer @chetmurthy pointed me a [fix on drjdn project](https://github.com/drjdn/p5scm/pull/4/files). I adapted the .opam file change to our `dune-project` quite easily. It was failing with:
```ocaml
File "bin/ged2gwb/ged2gwb.pp.ml", lines 533-539, characters 2-25:
533 | ..{Token.tok_func =
534 |     (fun s ->
535 |        make_date_lexing s,
536 |        {Plexing.Locations.locations = ref [| |]; overflow = ref true});
537 |    Token.tok_using = using_token; Token.tok_removing = (fun _ -> ());
538 |    Token.tok_match = tparse; Token.tok_text = (fun _ -> "<tok>");
539 |    Token.tok_comm = None}
Error: Some record fields are undefined: kwds
make: *** [Makefile:117: build] Error 1
```
I looked at the other modifications and tried a Token.kwds = something that was unbound! Mistral told me to set the field to a new empty hashtable `Hashtbl.create 10` (and that **we should replace it with an appropriate value for our use case!**).

Review heavily needed @ilankri @canonici please if you have time, it's important to fix and merge this as soon as possible. Thank you.

A warning remain since the beginning:
```ocaml
ocamlfind: [WARNING] Package `camlp5.extend': camlp5.extend SHOULD NOT be used with syntax camlp5o
```
Edit: removed § about “div Separated”… I did test on the wrong branch.. so waiting for CI here!

a2line committed:
prefer bold to white to improve readabilty when console background is white

chetmurthy commented:
Oh, I didn't realize you were using a custom Camlp5 lexer, too!  Yes, the fix is to provide a hashtable.  This change is also needed due to the need (at printer time) to know which are the keywords for the current syntax in use by Camlp5.  B/c the original and revised syntaxes have differing sets of keywords.

Here is an example of the change: https://github.com/camlp5/camlp5/blob/080b168b4461b0749f9b96bf9b1ca86d09776d6f/etc/pa_lispr.ml#L192

The idea that you need to supply a hashtable, which gets passed to the "tok_using" function (so it can add new tokens) and to the lexer function itself.

If this is confusing, I can help you make the change.

a2line commented:
Thank you @chetmurthy for your help and explanation even if I don’t get it fully. What I can tell: that ged2gwb code is mainly from Daniel (he’s obviously the only one who wrote things using his baby camlp4/5 here) and is in Geneweb since his first commit in the repository in September 1998!! There were tons of talks about getting rid of camlp5 syntax since 2014… and all of it was removed successivly except the one in this file. It is the last left, and I can’t remember why it wasn’t done, but quite recently (last January) someone at Ocamlpro did a work removing it for Geneweb v6.08 branch (was asked for Roglo database roglo.eu), but it’s still not done for the current v7.x master.
I think the Ocaml maintainers of Geneweb at Geneanet are able to understand all this. I also think my little line added should work, it was ok on Colab. I just wanted to activate the process a bit, because I saw the CI are not operational for their Gnt branch, and they probably would have missed it. And about the word ‘confusing’, I think it’s a bit light!! For this hashtable add, I have just listened to Mistral AI … very, very blindly! I was happy to see it does the job or seems to.

The irony is that it's now building on Unix and [failing on Windows](https://github.com/geneweb/geneweb/actions/runs/9232673644/job/25404155755) that don't have the good campl5 in it's repository: I'm stuck to 8.00.03 since a quite long time in my local cygwin installation, I'm even pinning it to avoid failing upgrades, and it's the same on Github CI. Should I add a condition in the dune.in if possible? I was tempted to report that tons of time. @chetmurthy Can I help in a way to have newer versions of camlp5 on Windows too?!

chetmurthy commented:
I have no idea how camlp5 gets to windows.  I assume you're using opam, so that means you're waiting for OPAM to push a new repository ?  No idea how that works, sorry.

Also, the changes in camlp5 8.03.00 are *breaking changes*.  So after you make these changes to geneweb, you will need to make a constraint of "camlp5 >= 8.03.00" in your OPAM file, or whatever build-system you use.

chetmurthy commented:
Hi, one last thing: since you mentioned that you had almost removed camlp5 from geneweb, I thought I would look to see what it was being used for in that one file (ged2gwb.ml).  It's being used to drive a parser for dates.  That's all (at least, that I can see).

Several thoughts:

1. you can probably rewrite that using Menhir, pretty easily.

2. But heck, I bet there are OCaml packages out there for date-parsing that do a perfectly fine job, heck, perhaps even better than the code you've got.

I'm not trying to drive you away from using Camlp5.  But if this is the *only* reason you're using it, well, I can't in good conscience not point out ways for you to migrate off.  Just a thought.  I'm happy to discuss further, and if I have misunderstood your use of Camlp5, I'm happy to investigate further: just point me at places where you use camlp5, and I'll  look further.

a2line commented:
I agree with you, of course we should remove camlp5 here! I said this so many times now… to every new dev coming here. 

For camlp5 under Windows as I'm using it since nearly 12 years. It now comes from what is called Sunset Opam repository, a fork from fdopen one that was depreciated. And that repository should be merged to the main Opam repository when Opam 2.2 will be out (in beta2.2 currently). As said the current version of camlp5 in this Windows Opam repo is 8.00.03, probably because nobody made a patch for Windows for upper versions, or needed them. Am I the only one?

https://discuss.ocaml.org/t/sunsetting-opam-repository-mingw/11632
https://github.com/ocaml-opam/opam-repository-mingw#sunset
https://github.com/ocaml-opam/opam-repository-mingw/tree/sunset/packages/camlp5

If you understand those patches and if perhaps they are easy to port to your newer versions. I can try them and we can PR this to this repository.

For the moment, I understand I need a condition to check for Unix or Win in dune file, can we do something like this
```
    (if 
        then (action (run camlp5o pr_o.cmo pa_extend.cmo q_MLast.cmo %{input-file}))
        else (action (run not-ocamlfind preprocess -package camlp5.extend,camlp5.quotations,camlp5.pr_o -syntax camlp5o %{input-file}))
    )
```
We definitly need a Windows executable for Geneweb.. we just released a beta 7.1 and wanted to get a 7.1 final release soon and now CI are broken… Tha's why I'm trying to find a compromise here. We can also pin camlp5 to 8.00.03 for everybody!

@hgouraud can you ask Ocamlpro if they can remove camlp5 from ged2gwd as it was done is 6.08 branch? It will solve everything…

chetmurthy commented:
The problem with pinning camlp5 to 8.00.03 is that for OCaml 5.2.0, you *must* use Camlp5 >= 8.03.00, b/c that's where the support is for OCaml 5.2.0 new features.

a2line commented:
I understand. But we are stuck at the moment to Ocaml 4.14.x see https://github.com/geneweb/geneweb/pull/1794 The new work from Ocaml pro uses no-naked-pointer Ocaml option+ancient dependancy to load database in RAM and it works on 4.13/4.14 only (not 5.x and not Windows at all…). Moreover for the same reason as for camlp5, Windows don't have access to Ocaml 5.xx in its current state of art (waiting for Opam 2.2 + the main repo to be updated…).

The older Ocamlpro PR removing camlp5 (1st commit) on the old 6.08 branch is here https://github.com/hgouraud/geneweb/pull/13 but I don't know if it's effective. Henri will tell us.

chetmurthy commented:
Oh ha, OK.  I get you now.  In that case, yeah, just pin camlp5 to an earlier version and be done with it.

a2line commented:
I hope we’ll get ancient to Ocaml 5.xx one day, never say never. And as things change very slowly here, I'll probably have time to test campl5 for Windows under Ocaml 5 one day, who knows…

So, yes, here are a few solutions for Geneweb:
* Remove the camlp5 syntax in `bin/ged2gwb/ged2gwb.ml` (will also save the build time of camlp5 when building from scratch).
* Pin camlp5 to 8.00.03 for all OS, that will probably work for Ocaml 4.08.1 to 5.1.1 (I can test easily this on PR 1794), note again that if we wan't -memory-in-cache in Unix releases we need to build with 4.14.x.
* eventually let Unix update camlp5 to latest so we can test Ocaml 5.2 that is coming soon (#veille technologique), and “carry” Windows (as always) finding a condition in the dune file? or dunno how… (see above, the run not-ocamlfind command is building ok for Unix, seems ok, before that Windows was ok).

a2line committed:
use not-ocamlfind only if camlp5 version >= 8.03.00

Add two new targets to patch/unpatch bin/ged2gwb/dune.in and ged2gwb.ml in
order that all various build work again. The perl regexps are compatible
between all OS (sed one are diverging between Unix and BSD). The modifications
should be transparent for every usages. The "distrib" target has been reworked
a bit to allow “make fmt distrib” used on Github Action.

a2line committed:
better format of commit message and distrib log, make gwd

add a new target “gwd” to build only gwd+gwc executables

ilankri commented:
I suggest we follow https://github.com/geneweb/geneweb/pull/1806#issuecomment-2130783765 to avoid all the _patch/unpatch machinery_ in the `Makefile`.

hgouraud commented:
Ideally, we should get rid of camlp5 altogether!!

a2line commented:
@ilankri: of course it's a far simpler solution that works but won't it be problematic to evolve later? That other solution  took me nearly 30 hours of multiple checks on all OS to have something stable and fully transparent, and it's working now. Henri uses it since a few days on his mac, this is harmless: The only problem is if you forget to commit your ged2gwb.ml before building… but we don't edit that file much, isn't it? It's sure it would have been better without it: It’s clearly a temporary solution to manage all possible campl5 version from ocaml 4.08 to 5.2 whatever the OS and installed camlp5 version. Of course pinning camlp5 to something < 8.03 works, but it's the easy solution (solution de facilité) and it does not work with Ocaml 5.2 at all, as said above: Ocaml 5.2 needs campl5 >= 8.03. Moreover as Henri said again the real solution is to remove camlp5 from ged2gwb.ml and fully get rid of the dependancy. I though it was a hard step but it seems it is used “for a unique date parser”… This is the real clean solution. If you can do this, this PR can be simplified a lot. Henri tried last week without success at the moment. Note that pinning ocamlformat to 0.24.1 is a pain for Windows because it don't build (was patched later). And that it don't work at all Ocaml5.2 (lot of work to be done, we have a pin that will be 3 releases late). Note that Ocaml 5.3 is coming this year that means we'll have 4 Ocaml version late… Pinning something should always be a temporary solution (same for calendars/unidecode.. we should “get further the pinning” and solve the problems instead).

pguyot commented:
Nice catch, however this PR adds `not-ocamlfind` as a requirement.

ilankri commented:
Merged as #1831.
